tests/test_irdl.py

```diff
@@ -0,0 +1,243 @@
+    eq_true_constraint = EqAttrConstraint(bool_true)
+    eq_true_constraint.verify(bool_true)
+
+
+def test_eq_attr_verify_wrong_parameters_fail():
+    """
+    Check that an EqAttrConstraint fails to verify an attribute with different
+    parameters.
+    """
+    bool_true = BoolData(True)
+    bool_false = BoolData(False)
+    eq_true_constraint = EqAttrConstraint(bool_true)
+    with pytest.raises(VerifyException) as e:
+        eq_true_constraint.verify(bool_false)
+    assert e.value.args[0] == (
+        f"Expected attribute {bool_true} but got {bool_false}")
+
+
+def test_eq_attr_verify_wrong_base_fail():
+    """
+    Check that an EqAttrConstraint fails to verify an attribute with a
+    different base attribute.
+    """
+    bool_true = BoolData(True)
+    int_zero = IntData(0)
+    eq_true_constraint = EqAttrConstraint(bool_true)
+    with pytest.raises(VerifyException) as e:
+        eq_true_constraint.verify(int_zero)
+    assert e.value.args[0] == (
+        f"Expected attribute {bool_true} but got {int_zero}")
+
+
+def test_base_attr_verify():
+    """
+    Check that a BaseAttr constraint verifies an attribute with the expected
+    base attribute.
+    """
+    eq_true_constraint = BaseAttr(BoolData)
+    eq_true_constraint.verify(BoolData(True))
+    eq_true_constraint.verify(BoolData(False))
+
+
+def test_base_attr_verify_wrong_base_fail():
+    """
+    Check that a BaseAttr constraint fails to verify an attribute with a
+    different base attribute.
+    """
+    eq_true_constraint = BaseAttr(BoolData)
+    int_zero = IntData(0)
+    with pytest.raises(VerifyException) as e:
+        eq_true_constraint.verify(int_zero)
+    assert e.value.args[0] == (
+        f"{int_zero} should be of base attribute {BoolData.name}")
+
+
+def test_any_attr_verify():
+    """Check that an AnyAttr verifies any attribute."""
+    any_constraint = AnyAttr()
+    any_constraint.verify(BoolData(True))
+    any_constraint.verify(BoolData(False))
+    any_constraint.verify(IntData(0))
+
+
+def test_anyof_verify():
+    """
+    Check that an AnyOf constraint verifies if one of the constraints
+    verify.
+    """
+    constraint = AnyOf([BaseAttr(BoolData), EqAttrConstraint(IntData(0))])
+    constraint.verify(BoolData(True))
+    constraint.verify(BoolData(False))
+    constraint.verify(IntData(0))
+
+
+def test_anyof_verify_fail():
+    """
+    Check that an AnyOf constraint fails to verify if none of the constraints
+    verify.
+    """
+    constraint = AnyOf([BaseAttr(BoolData), EqAttrConstraint(IntData(0))])
+    with pytest.raises(VerifyException) as e:
+        constraint.verify(IntData(1))
+    assert e.value.args[0] == f"Unexpected attribute {IntData(1)}"
+
+
+@dataclass
+class LessThan(AttrConstraint):
+    bound: int
+
+    def verify(self, attr: Attribute) -> None:
+        # This is just an assert because we expect this attribute to be an
+        # IntData.
+        assert isinstance(attr, IntData)
```

Comment:
Could this be written without this assertion? Maybe through the use of `None`? (Check for not None or something).

Reply:
Eheh, so this is related to testing tests.
Here, because this is in our test, we always expect this to be an `IntData`.
However, we never intend to test it without an `IntData`, because we don't care about testing that here.

Reply:
So I guess this is a good example on if we should check coverage of tests or not.
I changed this assert to an `if`, with an exception if the if fails.
However, now we don't cover this `if`, because we don't really want to write a test for that, that would make no sense.
So we don't have 100% coverage on that file, and we won't intend to, if we keep an `if` rather than the `assert`.
So I'm wondering, should we change it back to an assert, should we just don't care about this coverage number, or should we do something else?
I would suggest to keep the coverage, since it shows us what is covered and what is not covered, though I would not care that much about the number itself.

Reply:
My thought here was that you need to check this in order to ensure that the access goes to somehting of type int. However, everything can be None in Python, I was thinkg of check for `is_none` instead, such that you don't have to assert this here.

Maybe I am overlooking something though...

Reply:
Hmm, in typed Python this cannot be `None`, or at least shouldn't be, otherwise we would have `Attribute | None`.
So I'm actually not sure what you mean in your comment?

Reply:
Ah yes that makes a lot of sense. In typed python, it actually isn't the case that anything can be None. Or rather, it is the case but only because the typing is not really enforced. Forget my comment then 👍